core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1941,10 +1941,15 @@ class KafkaApis(val requestChannel: RequestChannel,
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
     val apiVersionRequest = request.body[ApiVersionsRequest]
+    val softwareName = apiVersionRequest.data.clientSoftwareName().split("-").last
+    var isXinfraClient = false
+    if (softwareName != null && softwareName.toLowerCase.contains("xinfra")) {
+      isXinfraClient = true;
+    }
```

Comment:
@groelofs  if we are worried this check/future observe might affect the performance, we can add another config to control this behavior for easy rollback. Not sure if that's needed though.

Reply:
The concern is that the pluggable observer interface hides the implementation from the code that uses it, so a naive change in the former made in ignorance could have a substantial effect on runtime performance, particularly at peak load. We're dealing with a two-day GCN right now that's fundamentally about an excessively hot request-handling path, so it's not just a theoretical worry.

That said, the hook implementation probably won't need to change in the future, so as long as the current implementation (in our internal MP, I assume?) is written with that in mind, it's probably OK to forego the config. Please include me on the latter PR, however.

Reply:
i'll add some comments about that. For now we're just storing the value in a cache and not performing any computations on it that should be expensive but i'll add that comment just in case